module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+			fe->fe_physical_len = BPE_GET_PSIZE(bp);
+			fe->fe_flags |= FIEMAP_EXTENT_DATA_INLINE |
+			    FIEMAP_EXTENT_NOT_ALIGNED;
+
+			if (BP_IS_ENCRYPTED(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_DATA_ENCRYPTED;
+			if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF)
+				fe->fe_flags |= FIEMAP_EXTENT_ENCODED;
+		} else {
+			if (i >= BP_GET_NDVAS(bp)) {
+				kmem_free(fe, sizeof (zfs_fiemap_entry_t));
+				continue;
+			}
+
+			if (BP_IS_ENCRYPTED(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_DATA_ENCRYPTED;
+			if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF)
+				fe->fe_flags |= FIEMAP_EXTENT_ENCODED;
+			if (BP_GET_DEDUP(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_SHARED;
+
+			/*
+			 * Report gang blocks as a single unknown extent.
+			 * Ideally we should be walking the gang block tree and
+			 * reporting all component-blocks as physical extents.
+			 */
+			if (BP_IS_GANG(bp)) {
+				fe->fe_flags |= FIEMAP_EXTENT_UNKNOWN;
+				fe->fe_physical_start = 0;
+				fe->fe_physical_len = 0;
+				fe->fe_vdev = 0;
+			} else {
+				fe->fe_physical_len = BP_GET_PSIZE(bp);
+
+				if (DVA_IS_VALID(&bp->blk_dva[i])) {
+					fe->fe_vdev =
+					    DVA_GET_VDEV(&bp->blk_dva[i]);
+					fe->fe_physical_start =
+					    DVA_GET_OFFSET(&bp->blk_dva[i]);
+				}
+			}
+
+			fe->fe_logical_len = BP_GET_LSIZE(bp);
+		}
+
+		/*
+		 * By default merge compatible adjacent block pointers in to a
+		 * single extent.  Embedded block pointers can never be merged.
+		 *
+		 * N.B. Block pointers provided by the iterator will always
+		 * be in logical offset order.  Therefore, it is sufficient
+		 * to check only the previously inserted entry when merging.
+		 */
+		pfe = avl_last(&fm->fm_extent_trees[i]);
+		if (pfe != NULL && !BP_IS_EMBEDDED(bp) &&
+		    !(fm->fm_flags & FIEMAP_FLAG_NOMERGE)) {
+			ASSERT3U(pfe->fe_logical_start + pfe->fe_logical_len,
+			    ==, fe->fe_logical_start);
+
+			if (BP_IS_HOLE(bp) && fe->fe_flags ==
+			    (pfe->fe_flags & ~FIEMAP_EXTENT_MERGED)) {
+				pfe->fe_logical_len += fe->fe_logical_len;
+				pfe->fe_flags |= FIEMAP_EXTENT_MERGED;
+				kmem_free(fe, sizeof (zfs_fiemap_entry_t));
+				continue;
+			}
+
+			if (!BP_IS_HOLE(bp) && fe->fe_flags ==
+			    (pfe->fe_flags & ~FIEMAP_EXTENT_MERGED) &&
+			    fe->fe_physical_start ==
+			    pfe->fe_physical_start + pfe->fe_physical_len &&
+			    fe->fe_vdev == pfe->fe_vdev) {
+				pfe->fe_logical_len += fe->fe_logical_len;
+				pfe->fe_physical_len += fe->fe_physical_len;
+				pfe->fe_flags |= FIEMAP_EXTENT_MERGED;
+				kmem_free(fe, sizeof (zfs_fiemap_entry_t));
+				continue;
+			}
+		}
+
+		/*
+		 * The FIEMAP documentation specifies that all encrypted
+		 * extents must also set the encoded flag.
+		 */
+		if (fe->fe_flags & FIEMAP_EXTENT_DATA_ENCRYPTED)
+			fe->fe_flags |= FIEMAP_EXTENT_ENCODED;
+
+		/*
+		 * Add the new extent to the copies tree.  This should never
+		 * conflict with an existing logical extent, but is handled
+		 * none the less by discarding the overlapping extent.
```

Comment:
Why?  If it's a bug for there to be an overlapping extent, seems we would wan to fail an assertion, rather than return incorrect info.

Reply:
If this relates to ditto blocks, Brian and I discussed this a bit.  Our conclusion was that it makes sense to return only a single copy for the file's logical extent, since most applications only care about whether blocks are allocated for that range or not.  It is possible to specify `FIEMAP_FLAG_COPIES` to return multiple blocks for a given logical extent.

Reply:
It was convenient when originally writing this code to not ASSERT.  I'll convert this to `avl_add()` to get rid of the dead code.